org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/BatchImageBuilder.java

```diff
@@ -24,15 +24,30 @@
 import org.eclipse.jdt.internal.core.util.Messages;
 import org.eclipse.jdt.internal.core.util.Util;
 
+import static org.eclipse.jdt.internal.core.JavaModelManager.trace;
+
 import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 @SuppressWarnings({ "rawtypes", "unchecked" })
 public class BatchImageBuilder extends AbstractImageBuilder {
 
-	IncrementalImageBuilder incrementalBuilder; // if annotations or secondary types have to be processed after the compile loop
-	ArrayList secondaryTypes; // qualified names for all secondary types found during batch compile
-	Set<String> typeLocatorsWithUndefinedTypes; // type locators for all source files with errors that may be caused by 'not found' secondary types
-	final CompilationGroup compilationGroup;
+	private IncrementalImageBuilder incrementalBuilder; // if annotations or secondary types have to be processed after the compile loop
+	private ArrayList secondaryTypes; // qualified names for all secondary types found during batch compile
+	private Set<String> typeLocatorsWithUndefinedTypes; // type locators for all source files with errors that may be caused by 'not found' secondary types
+	private final CompilationGroup compilationGroup;
+
+	/*  leave 2 threads for compiler + reader.*/
+	private static final ExecutorService WRITER_EXECUTOR = Executors.newFixedThreadPool(Math.max(1, Runtime.getRuntime().availableProcessors() - 2), r -> {
+				Thread t = new Thread(r, "Compiler Class File Writer"); //$NON-NLS-1$
+				t.setDaemon(true);
+				return t;
+			});
+	private final Map<IFile, byte[]> classContents = new HashMap<>();
+	private int classContentsBytesQueued = 0;
```

Comment:
Should it be long? In an int one can only have ~2 GB of content. Surely not usual, but there is no need to limit the compiler. E.g. we run our application with 31 GB heap on 256 GB workstations and workspaces of hundreds thousands of files, where 2 GB is nothing.

Small style: don't init to default zero please.

Reply:
i can change that. however i have hardcoded a 100MB limit.